### PR TITLE
Integrate StableLM (#254)

### LINKED ISSUE #254: [Proposal] Add Support for StableLM Base & Tuned, 3B & 7B
### Proposal 

Stability AI released instruction tuned (tuned) and pre-trained (base) models of size 3B and 7B, open source: https://huggingface.co/stabilityai/stablelm-tuned-alpha-7b (see others here https://huggingface.co/stabilityai/ ).

### Motivation

Work on instruction following (fine-tuned, not RLHF) models seems exciting to me. Also these models are trained on 800B tokens each which is great and they're probably much better to study than other similarly sized or larger models in TransformerLens. 

They use the NeoX architecture so should not be that hard to add in

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/Farama-Foundation/MiniGrid/issues) in the repo (**required**)

### PR DESCRIPTION
**Pull request by glerzing**
# Description

Enables to use Stable_LM, and adds a demo. The tuned version of Stable_LM is a bit special because, like ChatGPT, there are special tokens <|SYSTEM|>, <|USER|> and <|ASSISTANT|>. It is specially tuned with instruction fine-tuning for chat. These special tokens are also end tokens, so this needs a special treatment in the generate function, in order to stop when StableLM outputs <|USER|>.

Fixes #254 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

File: transformer_lens/HookedTransformer.py [hunk 1]
```diff
@@ -1470,20 +1471,39 @@ def generate(
...
                 self.cfg, self.cfg.device, batch_size
             )
         else:
             past_kv_cache = None
 
-        if stop_at_eos and eos_token_id is None:
-            assert (
+        stop_tokens = []
+        pad_token_id = 0
+        if stop_at_eos:
```
Comment (ArthurConmy) [inline transformer_lens/HookedTransformer.py:1485]:
```suggestion
        if stop_at_eos:
            pad_token_id = 0
```

Comment (ArthurConmy) [reply transformer_lens/HookedTransformer.py:1485]: I don't think we want to leave a variable in scope if it's not used at all. Let me know if the type checker isn't happy and we can find a workaround

File: transformer_lens/HookedTransformer.py [hunk 2]
```diff
@@ -1470,20 +1471,39 @@ def generate(
...
             past_kv_cache = HookedTransformerKeyValueCache.init_cache(
                 self.cfg, self.cfg.device, batch_size
             )
         else:
             past_kv_cache = None
 
-        if stop_at_eos and eos_token_id is None:
-            assert (
+        stop_tokens = []
+        pad_token_id = 0
```
Comment (ArthurConmy) [inline transformer_lens/HookedTransformer.py:1484]: Also, can well call this `eos_token_for_padding`? HuggingFace uses `pad_token_id` and we don't so reading this code was pretty confusing

File: transformer_lens/HookedTransformer.py [hunk 3]
```diff
@@ -1520,12 +1540,14 @@ def generate(
                 temperature=temperature,
                 freq_penalty=freq_penalty,
                 tokens=tokens,
-            ).to(devices.get_device_for_block_index(0, self.cfg))
+            ).to(device)
 
             if stop_at_eos:
                 # For all unfinished sequences, add on the next token. If a sequence finished, we throw away the generated token and instead add an EOS token to pad.
```
Comment (ArthurConmy) [inline transformer_lens/HookedTransformer.py:1546]:
```suggestion
                # For all unfinished sequences, add on the next token. If a sequence finished, we throw away the generated token and instead add the eos_for_padding token
```

File: demos/stable_lm.ipynb [hunk 4]
```diff
@@ -0,0 +1,5897 @@
...
+        "    import google.colab\n",
+        "    IN_COLAB = True\n",
+        "    print(\"Running as a Colab notebook\")\n",
+        "    %pip install git+https://github.com/neelnanda-io/TransformerLens.git\n",
+        "    %pip install circuitsvis\n",
+        "\n",
+        "    # PySvelte is an unmaintained visualization library, use it as a backup if circuitsvis isn't working\n",
+        "    # # Install another version of node that makes PySvelte work way faster\n",
+        "    # !curl -fsSL https://deb.nodesource.com/setup_16.x | sudo -E bash -; sudo apt-get install -y nodejs\n",
+        "    # %pip install git+https://github.com/neelnanda-io/PySvelte.git\n",
```
Comment (ArthurConmy) [inline demos/stable_lm.ipynb:5196]:
```suggestion
```

Remove this slower install

File: demos/stable_lm.ipynb [hunk 5]
```diff
@@ -0,0 +1,5897 @@
...
+        "    ipython.magic(\"load_ext autoreload\")\n",
+        "    ipython.magic(\"autoreload 2\")\n",
+        "\n",
+        "# Plotly needs a different renderer for VSCode/Notebooks vs Colab argh\n",
+        "import plotly.io as pio\n",
+        "if IN_COLAB or not DEVELOPMENT_MODE:\n",
+        "    pio.renderers.default = \"colab\"\n",
+        "else:\n",
+        "    pio.renderers.default = \"notebook_connected\"\n",
+        "print(f\"Using renderer: {pio.renderers.default}\")"
```
Comment (ArthurConmy) [inline demos/stable_lm.ipynb:5213]:
```suggestion
```

File: demos/stable_lm.ipynb [hunk 6]
```diff
@@ -0,0 +1,5897 @@
...
+      ]
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "import circuitsvis as cv\n",
```
Comment (ArthurConmy) [inline demos/stable_lm.ipynb:5219]:
```suggestion
```

**Review by ArthurConmy — CHANGES_REQUESTED** [review]
Cool stuff! Two minor changes to pad_token_id and some deletions and let's merge this!

File: transformer_lens/HookedTransformer.py (hunk 1 above)
Comment (glerzing) [reply transformer_lens/HookedTransformer.py:1485]: What you propose makes sense, but that leads to the warning `Local variable 'pad_token_id' might be referenced before assignment ` in PyCharm.

Comment (glerzing) [reply transformer_lens/HookedTransformer.py]: I took into account all your other comments.

**Review by ArthurConmy — APPROVED** [review]

**Pull request merged**